### PR TITLE
DD4hep: Filtered View Update

### PR DESCRIPTION
**Pull request by ianna**
#### PR description:

* Add a FilteredView constructor from a CompactView
* Add Shapes idetifies
* Allow white spaces in Vector of Strings
* Define an extra Trd1 constructor identical to a Trd2
* Add more unit tests

#### PR validation:

unit tests

#### if this PR is a backport please specify the original PR:

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
+1

File: DetectorDescription/DDCMS/src/DDFilteredView.cc [hunk 1]
```diff
@@ -239,6 +301,49 @@ vector<double> DDFilteredView::extractParameters() const {
...
+  return currVol.solid().dimensions();
+}
+
+const DDSolidShape DDFilteredView::shape() const {
+  return cms::dd::value(cms::DDSolidShapeMap, node_->GetVolume()->GetShape()->GetTitle());
+}
+
+template <>
+double DDFilteredView::get<double>(const char* key) const {
+  std::cout << "DDFilteredView::get<double> " << key << "\n";
```
Comment (kpedro88) [inline DetectorDescription/DDCMS/src/DDFilteredView.cc]: please remove all couts in this file

Comment (ianna) [reply DetectorDescription/DDCMS/src/DDFilteredView.cc]: thanks, done

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
+1

**Comment by kpedro88** [comment]
+1

**Comment by fabiocos** [comment]
@ianna @kpedro88 there is an Header inconsistency in this PR that appears in recent IBs:

```
>> Compiling src/DetectorDescription/DDCMS/interface/DDFilteredView.h 
src/DetectorDescription/DDCMS/interface/DDSolidShapes.h: In function 'std::__cxx11::string cms::dd::name(Mapping, V)':
src/DetectorDescription/DDCMS/interface/DDSolidShapes.h:17:40: error: return type 'std::__cxx11::string {aka class std::__cxx11::basic_string<char>}' is incomplete
     std::string name(Mapping a, V value) {
                                        ^
src/DetectorDescription/DDCMS/interface/DDSolidShapes.h:18:23: error: 'find_if' is not a member of 'std'
       auto pos = std::find_if(
                       ^~~~~~~
src/DetectorDescription/DDCMS/interface/DDSolidShapes.h:19:16: error: 'begin' is not a member of 'std'
           std::begin(a), std::end(a), [&value](const typename Mapping::value_type& t) { return (t.value == value); });
                ^~~~~
src/DetectorDescription/DDCMS/interface/DDSolidShapes.h:19:31: error: 'end' is not a member of 'std'
           std::begin(a), std::end(a), [&value](const typename Mapping::value_type& t) { return (t.value == value); });
                               ^~~
src/DetectorDescription/DDCMS/interface/DDSolidShapes.h:20:23: error: 'end' is not a member of 'std'
       if (pos != std::end(a)) {
                       ^~~
src/DetectorDescription/DDCMS/interface/DDSolidShapes.h:24:19: error: 'begin' is not a member of 'std'
       return std::begin(a)->name;
                   ^~~~~
src/DetectorDescription/DDCMS/interface/DDSolidShapes.h: In function 'typename Mapping::value_type::value_type cms::dd::value(Mapping, const string&)':
src/DetectorDescription/DDCMS/interface/DDSolidShapes.h:29:23: error: 'find_if' is not a member of 'std'
       auto pos = std::find_if(
                       ^~~~~~~
src/DetectorDescription/DDCMS/interface/DDSolidShapes.h:30:16: error: 'begin' is not a member of 'std'
           std::begin(a), std::end(a), [&name](const typename Mapping::value_type& t) { return (t.name == name); });
                ^~~~~
src/DetectorDescription/DDCMS/interface/DDSolidShapes.h:30:31: error: 'end' is not a member of 'std'
           std::begin(a), std::end(a), [&name](const typename Mapping::value_type& t) { return (t.name == name); });
                               ^~~
src/DetectorDescription/DDCMS/interface/DDSolidShapes.h:31:23: error: 'end' is not a member of 'std'
       if (pos != std::end(a)) {
                       ^~~
src/DetectorDescription/DDCMS/interface/DDSolidShapes.h:34:19: error: 'begin' is not a member of 'std'
       return std::begin(a)->value;
                   ^~~~~
src/DetectorDescription/DDCMS/interface/DDSolidShapes.h: At global scope:
src/DetectorDescription/DDCMS/interface/DDSolidShapes.h:59:14: error: 'array' in namespace 'std' does not name a template type
   const std::array<const cms::dd::NameValuePair<DDSolidShape>, 18> DDSolidShapeMap{
              ^~~~~
gmake: *** [config/SCRAM/GMake/Makefile.rules:1743: tmp/slc7_amd64_gcc700/src/DetectorDescription/DDCMS/src/DetectorDescriptionDDCMS/DDSolidShapes.h.chk_header] Error 1
```

To be fixed

File: DetectorDescription/DDCMS/interface/DDSolidShapes.h [hunk 2]
```diff
@@ -0,0 +1,80 @@
+#ifndef DETECTOR_DESCRIPTION_DDCMS_DD_SOLID_SHAPES_H
+#define DETECTOR_DESCRIPTION_DDCMS_DD_SOLID_SHAPES_H
+
+#include <iosfwd>
+
```
Comment (fabiocos) [inline DetectorDescription/DDCMS/interface/DDSolidShapes.h]: @ianna given the header problems I would say that here at least 

```
#include <array>
#include <algorithm>
```

are needed (to be verified), do you agree?

Comment (fabiocos) [reply DetectorDescription/DDCMS/interface/DDSolidShapes.h]: and 

```
#include <iterator>
```

as well

Comment (ianna) [reply DetectorDescription/DDCMS/interface/DDSolidShapes.h]: thanks, @fabiocos - I'll fix it in the next PR

**Pull request merged**